### DISCUSSION TITLE
Contact Birthdays calendar (3.7.0+)

### DISCUSSION THREAD

**Discussion by sprnza** [Kozea/Radicale#1463] [Ideas]
Is it me not able to find, or there is no an option to have contacts birthdays to be aggregated in a separate or already existed calendar?

**Comment by Unrud** [comment]
This is not supported on the server-side by Radicale. Some clients support it. (e.g. [Evolution](https://wiki.gnome.org/Apps/Evolution) or  [Birthday Adapter](https://f-droid.org/packages/org.birthdayadapter/))

**Comment by dirdi** [comment]
I understand that Radicale does not support this, yet. However, it would be convenient to have such a feature on the server-side, since many clients do only support the creation of birthday events through a plugin, resulting in one to install a ton of plugins on different clients.

Maybe this bug report could be reopened, as a feature request?

Btw. sabredav for example supports this since late 2015.

**Comment by kerbless** [comment]
Quoting @dirdi, is it hard to implement?

**Comment by reitzig** [comment]
FWIW, [Birthday Adapter](https://f-droid.org/packages/org.birthdayadapter/) doesn't seem to work on Android 14 (no updates since 2016) and Google Calendar does not show birthdays from contacts other than from Google accounts (doesn't work with DAVx5 contact accounts, anyway). So, an a current Android phone I'm sort of stuck short of installing a different calendar app ([aCalender](https://acalendar.tapirapps.de) might work, as per [davx5/faq](https://www.davx5.com/faq/contact-birthdays-in-calendars)).

Somehow, it seems _neater_ to have radicale expose a CalDAV calender based on a(ny) contacts store.

@Unrud, would you be open to discussing this as a feature request? And/or reviewing a PR?

**Comment by jlbribeiro** [comment]
@reitzig Regarding Android calendar apps, [Fossify Calendar](https://github.com/FossifyOrg/Calendar) has "Add contact birthdays" and "Add contact anniversaries" options; this only "displays them" in the app, it does not generate calendars available to other Android apps.

(Personally I coded something to auto-generate a birthday and anniversaries calendar based on my address book contacts, by leveraging the [`storage.hook`](https://radicale.org/v3.html#hook) mechanism, inspired by [this comment](https://github.com/Kozea/Radicale/issues/327#issue-113212486); this allows using those calendars both system-wide and across devices (which feels like a more "consistent" experience). Don't think it's worth sharing since it makes some assumptions on deployment/environment + it's coded in Go (and not Python), but I might consider it.)

Would rather have this as part of Radicale itself.

**Reply by theJoshMuller** [reply]
Agreed, I'd love to have this as part of Radicale.

**Reply by sprnza** [reply]
@jlbribeiro Could you please share the Go script if it's still around.

**Comment by pbiering** [comment]
To get this "magic" proper running I think there are two options:

1. regular job which scans address book and create/add/remove in a configured or predefined calendar related ics files in async way

2. doing this on request only on a virtual calendar (with caching files only) automatically created by option for each addressbook on the fly. On every trigger it checks content of related addressbook for any changes, update cache if needed and serve the result - no "physical" ICS files will be created.

**Reply by reitzig** [reply]
3. After initialization, hook into updates to the address book to update the birthday/anniversary calendar whenever a date is added/changed/removed.

**Comment by fin444** [comment]
I have created a simple script to handle this outside of Radicale using [VCF-to-ICS](https://github.com/FoxP/VCF-to-ICS/). Change the `CONTACTS_DIR` and `BIRTHDAYS_DIR`, and then set it to run daily with a cron or something, and it should Just Work. Though the birthday calendar will need to be separate with no other events, and be named "Birthdays".

```bash
CONTACTS_DIR=/var/lib/radicale/collections/collection-root/fin444/43a4b15d-2870-8ba2-b3d1-ef1c78b1410a
BIRTHDAYS_DIR=/var/lib/radicale/collections/collection-root/fin444/7137166a-0797-815c-8e2c-d287247c7011

shopt -s nullglob
rm "$BIRTHDAYS_DIR"/*.ics
for contact in "$CONTACTS_DIR"/*.vcf; do
	python vcf_to_ics.py -i "$contact" -o "$BIRTHDAYS_DIR/$(basename "$contact" .vcf).ics" -n Birthdays
done
```

**Reply by hadleyrich** [reply]
Nice solution. I was just contemplating writing a quick plugin to achieve this but something like this seems like it could be run from the storage hook in the config file from a quick look.

**Comment by superuser-does** [comment]
I found a [hook-based script on GitHub](https://github.com/iBigQ/radicale-birthday-calendar), though it unfortunately relies on Git, which may not be possible on shared hosting in particular. There are some alternative solutions [in its issues](https://github.com/iBigQ/radicale-birthday-calendar/issues/6). I agree a hook-based solution would be a good fit for Radicale. I think this could be deployed by-collection, but with shared collections coming soon in #1978, I'm really not sure what a simple interface would be to automatically generate birthday calendars (with the ability to opt some collections or users out, that is).

**Reply by pbiering** [reply]
It's planned for 3.8.0 (latest) to extend the new sharing implementation by a virtual "birthday" calendar overlay - technically the idea is introducting a new ShareType=bday, pointing with "PathMapped" to an addressbook collection and expose all entries on the fly and read-only as calendar collection named as "PathOrToken", which have a birthday entry active. Please give me some weeks...

**Reply by pbiering** [reply]
here please try-out: https://github.com/pbiering/Radicale/tree/sharing-bday

see also https://github.com/Kozea/Radicale/discussions/1978

It' on top of current upstream, WebUI support not extended so far for "bday" shares.
@maxberger - should I merge this also into upstream or postpone?

**Reply by maxberger** [reply]
I think the main question is: Does the current UI break? 
* If not, feel free to merge
* If yes, let me know and do not merge.

Theoretically the right filters should apply so that birthdays are "just not visible", but should not break any functionality. That said, there is always a chance that something is not working correctly.

All that said: I will take a look at the new functionality and see if I can provide a UI for that.

**Reply by maxberger** [reply]
While we're on the topic of "virtual" calendars which use transformation: I have an alternative approach for consideration.

We could say that map/key are types of sharing, and that there is a  another property which defines transformations, with the default transformer being: no transformation.  A transformer could then be the "bday" transformer, which does the functionality described here. This would allow both key and map sharing of birthday calendars.

But it would also open up the possibility to add another transformer, a ¨free/busy" transformer. this would change a regular calender in a VFREEBUSY -> And for this one it would be great if they could be shared both as key or as map.

**Reply by maxberger** [reply]
> I think the main question is: Does the current UI break?
> 
>     * If not, feel free to merge
> 
>     * If yes, let me know and do not merge.
> 
> 
> Theoretically the right filters should apply so that birthdays are "just not visible", but should not break any functionality. That said, there is always a chance that something is not working correctly.
> 
> All that said: I will take a look at the new functionality and see if I can provide a UI for that.

Thinking about this some more: I will wait on an answer on the transformer model before creating a dedicated UI, since both would be reflected differently.

**Reply by pbiering** [reply]
"bday" implementation will not break current API or WebUI, it is simply an extension with it's own permission set and a different ShareType. Using "map" and a transformer flag is imho not that easy as "bday" has to convert VADDRESSBOOK into VCALENDAR (filtered by vcf having bday). A free-busy converter would stay inside same collection type VCALENDAR...if such is needed (later) one can think about how to implement this.

**Reply by pbiering** [reply]
WebUI related an additional ShareType only for VCALENDAR need to be added (if feature is enabled and permitted) - adding after "sharing by map" e.g. as "sharing as bday calendar" (even to same user possible).

**Reply by pbiering** [reply]
Regardig "bday" and permissions: this special map is always read-only (as it scans all vcards and convert to vevents on the fly), only the permission to enable properties overlay for the collection can be toggled.

**Reply by maxberger** [reply]
> WebUI related an additional ShareType only for VCALENDAR need to be added (if feature is enabled and permitted) - adding after "sharing by map" e.g. as "sharing as bday calendar" (even to same user possible).

Ok, I will make this a 3rd list, only to be visible for the cases you mentioned

> Regardig "bday" and permissions: this special map is always read-only (as it scans all vcards and convert to vevents on the fly), only the permission to enable properties overlay for the collection can be toggled.

Ok; I see that I eventually need to implement property overrides for users. I'll leave that out for the initial solution though for simplicity.

**Comment by pbiering** [comment]
Work done, please test latest upstream from now, container-latest will be updated on next daily run.

**Reply by pbiering** [reply]
3.7.0 is now available as pre-release

**Comment by ape364** [comment]
how to use it? how to create birthday calendar?

**Reply by pbiering** [reply]
You need to create a self-shared map of your addressbook collection using "bday" conversion, by API see here: https://github.com/Kozea/Radicale/blob/master/SHARING.md#api-hook-mapcreate - for the WebUI there is no documentation with screenshots or workflow sequences so far available.

**Reply by schildbach** [reply]
Has anyone managed to use this for a birthday calender and post the exact steps that need to be taken?

**Reply by pbiering** [reply]
> Has anyone managed to use this for a birthday calender and post the exact steps that need to be taken?

Via API: https://github.com/Kozea/Radicale/blob/master/SHARING.md#virtual-bday-collection

WebUI doc is still missing, see https://github.com/Kozea/Radicale/issues/2085

**Reply by schildbach** [reply]
I tried the linked instructions, but I always get

`The requested resource could not be found.`

> WebUI doc is still missing

Where is the WebUI for this feature? With each release I explore the entire UI for if birthdays have been added but I cannot find it.

**Reply by maxberger** [reply]
Dear @schildbach ,

Proper documentation is still pending https://github.com/Kozea/Radicale/issues/2085

For now, you need to create a calendar, and when you share it you have bday as one of the sharing options. You can also share with yourself.

Max

**Reply by schildbach** [reply]
Can I share calendars via UI? I see an URL of the calendar that I can copy, do you mean that by "sharing"?

**Reply by pbiering** [reply]
Yes, it's possible by UI with some prerequirements, via CLI it's documented here:

https://github.com/Kozea/Radicale/blob/master/SHARING.md#virtual-bday-collection

**Reply by kaulin** [reply]
> I tried the linked instructions, but I always get
> 
> `The requested resource could not be found.`

Late response to this, but having just finished setting up bday calendar successfully, I just want to ask if you used the href of the address book, not just its name in: "PathMapped=/owner/addressbook/"?

And you should have a sharing block in your config, with something similar to this:
[sharing]
type = files
collection_by_map = True
permit_create_map = True

PS. Big thank you for implementing this feature. I was making a migration to radicale a while back, but was never satisfied with a scripted solution. One question though: if the conversion template is modified later on, what is the best way to implement those changes on existing calendar entires? Removing cache files works, but seems a bit hacky.

**Reply by pbiering** [reply]
> if the conversion template is modified later on, what is the best way to implement those changes on existing calendar entires? Removing cache files works, but seems a bit hacky.

The virtual collection itself has no cache, content is generated on the fly. On next REPORT request the result should be modified and client should display new format.

**Comment by pbiering** [comment]
Implemented 3.7.0+

**Comment by jasper-r75** [comment]
For people who have a n8n instance. I have created a workflow that works for me.. It syncs birthdays in your contacts daily, to a new "Birthday" calendar. You have to create that calendar manually first, in the Radicale web interface. 

Mind you, I'm more of a tinkerer than a programmer, and asked for some help of Gemini.

Just open or download the Template.json file.  Copy the json and paste it (Ctrl+V) onto a new, empty workflow, and follow the sticky note instructions to fill in your addressbook and birthday calendar urls and create your credentials.

https://github.com/jasper-r75/n8n-Radicale-CardDAV-Birthdays-to-CalDAV-Sync-Template-/tree/main
